### PR TITLE
Cstr verify safety contracts of unsafe functions strlen, from_bytes_with_nul_unchecked 

### PR DESCRIPTION
**Pull request by rajathkotyal**
Towards #150

Annotates and verifies the safety contracts for the following unsafe functions:

`from_bytes_with_nul_uncheked` `core::ffi::c_str`
`strlen` `core::ffi::c_str`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [inline library/core/src/ffi/c_str.rs:984]: Relevant code below this

File: library/core/src/ffi/c_str.rs
Comment (Yenyun035) [inline library/core/src/ffi/c_str.rs:1031]: How about `kani::any_where(|x| *x >= 0 && x < MAX_SIZE)`?

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [inline library/core/src/ffi/c_str.rs:883]: The stubbing part is a bit unclear : 

Yeah. Since Kani doesnt support external runtime functions, I tried stubbing them, but the functions are in a nested manner (the extern "C" block) , we cannot stub nested functions since its not in the scope and is invoked during runtime. We could try to explicitly mock this function but Im not sure if that is expected.

i.e. the extern "C" block is taken out of this function maintaining functionality of the code OR, we mock the working of the strlen. Reference : https://model-checking.github.io/kani/rfc/rfcs/0002-function-stubbing.html
```
#[inline]
#[unstable(feature = "cstr_internals", issue = "none")]
#[cfg_attr(bootstrap, rustc_const_stable(feature = "const_cstr_from_ptr", since = "1.81.0"))]
#[rustc_allow_const_fn_unstable(const_eval_select)]
const unsafe fn strlen(ptr: *const c_char) -> usize {
    const fn strlen_ct(s: *const c_char) -> usize {
        let mut len = 0;

        // SAFETY: Outer caller has provided a pointer to a valid C string.
        while unsafe { *s.add(len) } != 0 {
            len += 1;
        }

        len
    }

    #[inline]
    fn strlen_rt(s: *const c_char) -> usize {
        extern "C" {
            /// Provided by libc or compiler_builtins.
            fn strlen(s: *const c_char) -> usize;
        }

        // SAFETY: Outer caller has provided a pointer to a valid C string.
        unsafe { strlen(s) }
    }

    intrinsics::const_eval_select((ptr,), strlen_ct, strlen_rt)
}
```

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [inline library/core/src/ffi/c_str.rs:893]: can we stub only strlen_rt here?

File: library/core/src/ffi/c_str.rs
Comment (Yenyun035) [reply library/core/src/ffi/c_str.rs:890]: @rajathkotyal FYI, strlen was updated:
```rust
const unsafe fn strlen(ptr: *const c_char) -> usize {
    const_eval_select!(
        @capture { s: *const c_char = ptr } -> usize:
        if const {
            let mut len = 0;

            // SAFETY: Outer caller has provided a pointer to a valid C string.
            while unsafe { *s.add(len) } != 0 {
                len += 1;
            }

            len
        } else {
            extern "C" {
                /// Provided by libc or compiler_builtins.
                fn strlen(s: *const c_char) -> usize;
            }

            // SAFETY: Outer caller has provided a pointer to a valid C string.
            unsafe { strlen(s) }
        }
    )
}
```

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:917]: Thank you, this resolves the nested module blocker

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [inline library/core/src/ffi/c_str.rs:997]: Since strlen isnt public, I wasnt able to access it in the verify module. Figured of verifying using the function below. It works but Im not sure if this is expected.

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [inline library/core/src/ffi/c_str.rs:1029]: Based on my research on stubs : 
https://github.com/model-checking/kani/issues/2423
https://github.com/firecracker-microvm/firecracker/blob/dbd9a84b11a63b5e5bf201e244fe83f0bc76792a/src/rate_limiter/src/lib.rs#L607-L639
https://github.com/model-checking/kani/pull/2658

I figured we can stub the external strlen function individually and write a proof for it, But this seems to not work as expected and gives me this error.

```
   Compiling dummy v0.1.0 (/Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/target/kani_verify_std)
error: Failed to find contract closure `__kani_recursion_check_strlen` in function `ffi::c_str::strlen`

error: could not compile `core` (lib) due to 1 previous error
error: Failed to execute cargo (exit status: 101). Found 1 compilation errors.
```

File: library/core/src/ffi/c_str.rs
Comment (celinval) [reply library/core/src/ffi/c_str.rs:997]: Where did you place the verify module? You should be able to wrap this function with the verify module. A private function is still available for all descendants of the module where it is defined.

What error were you getting?

File: library/core/src/ffi/c_str.rs
Comment (celinval) [reply library/core/src/ffi/c_str.rs:1029]: The problem here is that the strlen you are stubbing is the one you are trying to check.

There's no way for Kani to stub a nested function. Can you pass `-Z c-ffi` to Kani and remove the stub to see if that would work? Otherwise, we can add support for strlen to Kani. I believe CBMC already has a model available.

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:997]: My bad, needed to invoke using the super keyword. Its working now. Thanks!

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs]: Hi @celinval , thank you, using the flag seems to work, I have made some changes to the code and pushed it. Pl review

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [inline library/core/src/ffi/c_str.rs:1001]: We are using proof_for_contract and wouldn't want this assumption of checking nul before last byte be in our harness, But added here to improve brevity. Can move the pre condition check to the function definition requires clause as well. Something like : 

```
#[requires(!ptr.is_null() && {
    let mut found_null = false;
    let mut valid = true;
    for i in 0..isize::MAX as usize {
        if unsafe { *ptr.add(i) } == 0 {
            found_null = true;
            break;
        } else {
            // Ensure no interior null bytes
            if unsafe { *ptr.add(i) } == 0 {
                valid = false;
                break;
            }
        }
    }
    found_null && valid
})]
```

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:1001]: Update :  I tried the above method of including the precondition in the requires clause and received this error : 
```

VERIFICATION RESULT:
 ** 1 of 242 failed
Failed Checks: This is a placeholder message; Kani doesn't support message formatted at runtime
 File: "/Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/library/core/src/panicking.rs", line 418, in panicking::assert_failed_inner

VERIFICATION:- FAILED
Verification Time: 8.302092s

Summary:
Verification failed for - ffi::c_str::verify::check_strlen_contract
Complete - 0 successfully verified harnesses, 1 failures, 1 total.
```

Comment (tautschnig) [reply library/core/src/ffi/c_str.rs:1001]: Can you run Kani with the additional options `--harness check_strlen_contract --output-format old --enable-unstable --no-reachability-checks --cbmc-args --trace` to find the source line of the assertion that is failing?

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:1001]: I tried this command :

``` 
./scripts/run-kani.sh --kani-args --enable-unstable --harness ffi::c_str::verify --output-format old --no-assertion-reach-checks --cbmc-args --trace
```
 
but cbmc seems to be interfering with "--enable-unstable" flag and giving me  
``` 
Unknown option: --enable-unstable
Summary:
Verification failed for - ffi::c_str::verify::check_strlen_contract
Complete - 0 successfully verified harnesses, 1 failures, 1 total.
```

Not sure if my command format is wrong

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:1011]: Checked the logs, It was a `ptr` dereferencing error where since we weren't checking for interior nul bytes before calling the `strlen` func, it was panicking since it encountered nul bytes. Figured we need to keep the assumption before passing to the function as well as in the requires clause. Pushed a new commit now, it is working.

File: library/core/src/ffi/c_str.rs
Comment (celinval) [inline library/core/src/ffi/c_str.rs:781]: What about something like this?
```rust
let mut next = ptr;
loop {
  if !can_dereference(next) {
    break false;
  } else if unsafe { *next == 0 } {
    break true;
  }
  next = next.wrapping_add(1);
}
```

File: library/core/src/ffi/c_str.rs
Comment (celinval) [inline library/core/src/ffi/c_str.rs:795]: I think you can simplify this a bit since we mostly care about safety. You could use `[slice_from_raw_parts](https://doc.rust-lang.org/std/ptr/fn.slice_from_raw_parts.html)` to create a slice ptr, check that we can dereference it, and ensure the last byte is 0.

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs]: Thanks. Just updated the code.

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs]: Sure. Updated the code.

File: library/core/src/ffi/c_str.rs
Comment (tautschnig) [inline library/core/src/ffi/c_str.rs:1022]: In the precondition you use `contains` while here you use (both for the assumption and the assertion) explicit loops. I am wondering whether there is a performance difference between those approaches? Would you mind experimenting with 1) consistently using `contains` and 2) consistently using loops?

**Review by tautschnig — APPROVED** [review]

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:1022]: postcondition you mean?

Comment (tautschnig) [reply library/core/src/ffi/c_str.rs:1022]: Actually true of both: you also have it in a precondition in line 439.

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:1022]: Oh gotcha. That was for `from_bytes_with_nul_unchecked` . Will check and update

Comment (tautschnig) [reply library/core/src/ffi/c_str.rs:1022]: Thank you for gathering this data! I'd suggest you use a uniform solution, be it `contains` or loops. I'll leave it to you to pick.

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:1004]: FYI : These were the error I encountered  : 

```
Violated property:
  file /Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/library/core/src/panicking.rs function panicking::assert_failed_inner line 418 thread 0
  This is a placeholder message; Kani doesn't support message formatted at runtime
  FALSE

[panicking::assert_failed_inner.assertion.2] line 418 This is a placeholder message; Kani doesn't support message formatted at runtime: FAILURE

/Users/blaze/Documents/3rd_sem/practicum/verify-rust-std/library/core/src/panicking.rs function panicking::panic_nounwind
```

File: library/core/src/ffi/c_str.rs
Comment (celinval) [inline library/core/src/ffi/c_str.rs:768]: Once your other PR merges, can you please update this to use the new function?

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:768]: Sure will do

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:787]: Do you mean we remove this entirely or either keep this or the assertions in the code? Im confused how we are going to validate the results if we remove the ensures caluse

File: library/core/src/ffi/c_str.rs
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:1025]: Okay. Just a quick check, do you mean to say this code below is a sufficient amount of check for strlen? @celinval 

conditions : 
```
#[requires(!ptr.is_null() && is_null_terminated(ptr))]
#[ensures(|&result| result < isize::MAX as usize)]
```

proof  harness :  
```
    #[kani::proof_for_contract(super::strlen)]
    #[kani::unwind(33)]
    fn check_strlen_contract() {
        const MAX_SIZE: usize = 32;
        let string: [u8; MAX_SIZE] = kani::any();
        
        let ptr = string.as_ptr() as *const c_char;
        
        unsafe {
            let result_len = super::strlen(ptr);
            assert_eq!(string[result_len], 0);
        }
    }
```

Note : this does successfully pass.
```
Checking harness ffi::c_str::verify::check_strlen_contract...

VERIFICATION RESULT:
 ** 0 of 117 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 5.4615207s

Complete - 1 successfully verified harnesses, 0 failures, 1 total.
```

Comment (Yenyun035) [reply library/core/src/ffi/c_str.rs:1025]: I think this one doesn't check for all string <= max_size but only the string == max_size.

File: library/core/src/ffi/c_str.rs
Comment (celinval) [reply library/core/src/ffi/c_str.rs:787]: Since we've been looking at safety verification at this point, we need to ensure no UB occurs and that all objects that may be modified by the function are left in a safe state.

Further checks are nice to have but not necessary. If we do add functional post-conditions, I would prefer a simpler ensures clause so it's not hard to maintain

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs]: Simplified code by retaining essential checks and removed redundant verifications from the proof harness, pl review @celinval

**Comment by rajathkotyal** [comment]
Hi @celinval , incorporated all changes and removed redundant checks, could you review once again and approve the workflow if everything checks out? Let me know if there is anything missing. Thanks!

**Comment by Yenyun035** [comment]
The workflow failed likely due to the following error. I ran on my Macbook Pro (Apple M1 Pro, 16 GB RAM, Sequoia 15.0.1)
```
Checking harness ffi::c_str::verify::check_strlen_contract...

VERIFICATION RESULT:
 ** 1 of 158 failed (157 undetermined)
Failed Checks: call to foreign "C" function `strlen` is not currently supported by Kani. Please post your example at https://github.com/model-checking/kani/issues/2423
 File: "/Users/yew005/Docs/Academic/CMU/Fall24/practicum/vrs/library/core/src/ffi/c_str.rs", line 798, in ffi::c_str::strlen::{closure#2}::{closure#1}::runtime

VERIFICATION:- FAILED
** WARNING: A Rust construct that is not currently supported by Kani was found to be reachable. Check the results for more details.
Verification Time: 1.4835527s

Summary:
Verification failed for - ffi::c_str::verify::check_strlen_contract
Complete - 0 successfully verified harnesses, 1 failures, 1 total.
```

**Comment by celinval** [comment]
> The workflow failed likely due to the following error. I ran on my Macbook Pro (Apple M1 Pro, 16 GB RAM, Sequoia 15.0.1)
> 
> ```
> Checking harness ffi::c_str::verify::check_strlen_contract...
> 
> VERIFICATION RESULT:
>  ** 1 of 158 failed (157 undetermined)
> Failed Checks: call to foreign "C" function `strlen` is not currently supported by Kani. Please post your example at https://github.com/model-checking/kani/issues/2423
>  File: "/Users/yew005/Docs/Academic/CMU/Fall24/practicum/vrs/library/core/src/ffi/c_str.rs", line 798, in ffi::c_str::strlen::{closure#2}::{closure#1}::runtime
> 
> VERIFICATION:- FAILED
> ** WARNING: A Rust construct that is not currently supported by Kani was found to be reachable. Check the results for more details.
> Verification Time: 1.4835527s
> 
> Summary:
> Verification failed for - ffi::c_str::verify::check_strlen_contract
> Complete - 0 successfully verified harnesses, 1 failures, 1 total.
> ```

That means we need to enable C-FFI support like #204

**Comment by Yenyun035** [comment]
> That means we need to enable C-FFI support like https://github.com/model-checking/verify-rust-std/pull/204
@celinval Got it. Thank you for mentioning that. This means #204 needs to be merged first.

**Comment by carolynzech** [comment]
@Yenyun035 can you fix the failing Rust tests and then we'll take a look if there's still issues?

**Comment by Yenyun035** [comment]
@carolynzech Sure. The failed Rust tests were due to unused `is_null_terminated` helper function we created for verifying `strlen`. Is it better to annotate it with `#[cfg(kani)]` or `allow_unused_import`?

**Comment by Yenyun035** [comment]
@carolynzech Sure. I annotated `is_null_terminated` helper function with `#[cfg(kani)]` now. I think it should fix the failed rust tests. If not, I'll try to figure out another solution.

File: library/core/src/ffi/c_str.rs
Comment (carolynzech) [inline library/core/src/ffi/c_str.rs:14]:
```suggestion
#[allow(unused_imports)]
use crate::ub_checks::can_dereference;
```

Comment (Yenyun035) [reply library/core/src/ffi/c_str.rs:15]: Thank you.

**Comment by Yenyun035** [comment]
@tautschnig Fixed. Thank you.

**Review by tautschnig — APPROVED** [review]

**Review by carolynzech — APPROVED** [review]
Nice!

**Comment by rajathkotyal** [comment]
Thanks @Yenyun035 !

**Pull request merged**
